flake.nix

```diff
@@ -0,0 +1,448 @@
+{
+  description = "BDK Flake to run all tests locally and in CI";
+
+  inputs = {
+    # stable nixpkgs
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.05";
+    # pin bitcoind to a specific version
+    # find instructions here:
+    # <https://lazamar.co.uk/nix-versions>
+    # pinned to 0.25.0
+    nixpkgs-bitcoind.url = "github:nixos/nixpkgs?rev=9957cd48326fe8dbd52fdc50dd2502307f188b0d";
+    # Blockstream's esplora
+    # inspired by fedimint CI
+    nixpkgs-kitman.url = "github:jkitman/nixpkgs/add-esplora-pkg";
+    crane = {
+      url = "github:ipetkov/crane";
+      inputs.nixpkgs.follows = "nixpkgs";
+    };
+    rust-overlay = {
+      url = "github:oxalica/rust-overlay";
+      inputs = {
+        nixpkgs.follows = "nixpkgs";
+        flake-utils.follows = "flake-utils";
+      };
+    };
+    flake-utils.url = "github:numtide/flake-utils";
+    advisory-db = {
+      url = "github:rustsec/advisory-db";
+      flake = false;
+    };
+    pre-commit-hooks.url = "github:cachix/pre-commit-hooks.nix";
+  };
+
+  outputs = { self, nixpkgs, nixpkgs-bitcoind, nixpkgs-kitman, crane, rust-overlay, flake-utils, advisory-db, pre-commit-hooks, ... }:
+    flake-utils.lib.eachDefaultSystem (system:
+      let
+        overlays = [ (import rust-overlay) ];
+        lib = pkgs.lib;
+        stdenv = pkgs.stdenv;
+        isDarwin = stdenv.isDarwin;
+        libsDarwin = with pkgs.darwin.apple_sdk.frameworks; [
+          # Additional darwin specific inputs can be set here
+          Security
+          SystemConfiguration
+          CoreServices
+        ];
+
+        pkgs = import nixpkgs {
+          inherit system overlays;
+        };
+        pkgs-bitcoind = import nixpkgs-bitcoind {
+          inherit system overlays;
+        };
+        pkgs-kitman = import nixpkgs-kitman {
+          inherit system;
+        };
+
+        # Signed Commits
+        signed-commits = pkgs.writeShellApplication {
+          name = "signed-commits";
+          runtimeInputs = [ pkgs.git ];
+          text = ''
+            # Function to verify the signature of the last commit
+            verify_signature() {
+            	local commit_hash
+              commit_hash=$(git rev-parse HEAD)
+            	if ! git verify-commit "$commit_hash"; then
+            		echo "Error: Last commit ($commit_hash) is not signed."
+            		exit 1
+            	fi
+            }
+
+            # Verify the signature of the last commit
+            verify_signature
+
+            # Allow the push to proceed if the signature is valid
+            exit 0
+          '';
+        };
+
+        # Toolchains
+        # latest stable
+        rustTarget = pkgs.rust-bin.stable.latest.default;
+        # we pin clippy instead of using "stable" so that our CI doesn't break
+        # at each new cargo release
+        rustClippyTarget = pkgs.rust-bin.stable."1.67.0".default;
```

Comment:
**thought:** Since everything is pinned, splitting between "latest and clippy" seems unnecessary and can have a cost of number of toolchains that need to be downloaded.

Reply:
I don't think so. We have 3 different Rust versions:

- latest
- clippy, pinned to 1.67.0
- MSRV, pinned to 1.57.0